### PR TITLE
Uplift `clippy::precedence` lint

### LINKED ISSUE #117155: "-NUM.method()" is parsed in surprising ways
This came up [here](https://github.com/rust-lang/rust/issues/99842#issuecomment-1673625133) but I am pretty sure I ran into this before: Rust parses `-5.5f32.method()` as `-(5.5f32.method())`. I don't know if that is the parsing result that anyone expects, I certainly would never expect it `-5.5f32` is a float literals, and that is what I want to apply the method to, I never even considered that the compiler might interpret this as a unary minus applied to the method result.

Assuming that we cannot change the parser, I think we should spend some effort defending against the possible confusion here. For instance:
- lint against `-5.5f32.method()` (suggesting to write ``-(5.5f32.method())`` if that's really what one meant)
- have rustfmt add the parentheses that make precedence explicit

### PR DESCRIPTION
**Pull request by Urgau**
This PR aims at uplifting the `clippy::precedence` lint into rustc.

## `ambiguous_precedence`

(warn-by-default)

The `ambiguous_precedence` lint checks for operations where precedence may be unclear and suggests adding parentheses.

### Example

```rust
1 << 2 + 3; // equals 32, while `(1 << 2) + 3` equals 7
-1i32.abs(); // equals -1, while `(-1i32).abs()` equals 1
```

### Explanation

Unary operations take precedence on binary operations and method calls take precedence over unary precedence these precedence may be unexpected without parentheses.
Setting the precedence explicitly makes the code clearer and avoid potential bugs.

-----

As implemented in this PR, the lint **only** lints on
 - a binary operation (`<<`, `>>`...) where either the left and/or right expression is a non-bitwise operation (`*`, `+`, `/`...)
 - or when a literal number is preceded by a `-` (minus operator) and followed by some method calls

-----

Fixes https://github.com/rust-lang/rust/issues/117155 (cc @RalfJung)
@rustbot labels +I-lang-nominated
r? compiler

### CODE REVIEW

File: compiler/rustc_lint/src/precedence.rs [hunk 1]
```diff
@@ -0,0 +1,132 @@
+use rustc_ast::token::LitKind;
+use rustc_ast::{BinOpKind, Expr, ExprKind, MethodCall, UnOp};
+use rustc_span::source_map::Spanned;
+
+use crate::lints::{PrecedenceDiag, PrecedenceUnarySuggestion, PrecedenceUnwarySuggestion};
+use crate::{EarlyContext, EarlyLintPass, LintContext};
+
+// List of functions `f(x)` where `f(-x)=-f(x)` so the
+// precedence doens't matter.
+const ALLOWED_ODD_FUNCTIONS: [&str; 14] = [
```
Comment (RalfJung) [inline compiler/rustc_lint/src/precedence.rs:10]: Personally I'd prefer it to warn consistently, even if the semantics of the function mean it doesn't matter. (a) that seems more consistent and less surprising, and (b) I am writing tests which ensure that `sin` behaves correctly, and I care about which sign the argument actually has.

**Comment by asquared31415** [comment]
Should unary `!` be linted on in the same way that unary `-` is?
```rust
fn main() {
    let x = !10_i32.abs();
    println!("{:#034b}", x);
    let x = (!10_i32).abs();
    println!("{:#034b}", x);
}
```
```
0b11111111111111111111111111110101
0b00000000000000000000000000001011
```
I'm not sure what most people's expectations are of `!`, I could see it being more clear that it's an expression as opposed to `-` which some people might think is part of the literal.

File: compiler/rustc_lint/src/precedence.rs (hunk 1 above)
Comment (Urgau) [reply compiler/rustc_lint/src/precedence.rs]: Not warn on odd method calls was asked in https://github.com/rust-lang/rust-clippy/issues/4892, except for the obvious "the precedence doesn't matter" with them, no other reasons were given, but the consistency issue seems higher to me.

Anyway I agree, this behavior is unexpected and weird, I've removed this logic, we can always re-add it later if people really want it.

**Comment by traviscross** [comment]
@Urgau:

Regarding the nomination, the question that will come up in the T-lang call is whether or not this is an appropriate lint for **`rustc`**.  That is, this may be a reasonable lint in general, but for this to land in `rustc`, we need to explain why this one specifically makes sense to uplift.  It'd be worth leaving a comment to discuss, e.g., other comparable lints we have in `rustc` and how this one fits in with those conceptually.

**Comment by Urgau** [comment]
> Should unary `!` be linted on in the same way that unary `-` is?

I'm not sure, I've implemented it locally, but almost every time it was triggered, I've had the impression that it was a false positive. So I don't think so, at least for the time being.

> Regarding the nomination, the question that will come up in the T-lang call is whether or not this is an appropriate lint for **`rustc`**...

I don't have much to say, other than clippy's `precedence` lint would have catch the ambiguity in https://github.com/rust-lang/rust/issues/117155 and that we don't currently have anything in `rustc` that resemble it currently, but we however have other lints about ambiguity: [`ambiguous_associated_items`](https://doc.rust-lang.org/rustc/lints/listing/deny-by-default.html#ambiguous-associated-items) or [`ambiguous_glob_reexports`](https://doc.rust-lang.org/rustc/lints/listing/warn-by-default.html#ambiguous-glob-reexports).

**Comment by camsteffen** [comment]
This lint is a bit slippery to define because there is a sliding scale of how much a human might know the precedence rules. A lesser experienced programmer might prefer parenthesis to clarify precedence of `||` vs. `&&` while a highly experienced programmer might prefer to never have parenthesis that are strictly not necessary. I do think we can define it a bit more narrowly if we say that it is about "uncommon operators" such as `|` or `<<`. So I might name this lint `uncommon_operator_precedence`. Of course there is still some subjectivity about which operators are uncommon, but we can all probably agree that `&&` and `||` are _not_ uncommon.

**Comment by Urgau** [comment]
To be clear this lint as implemented in this PR **only** lints on:
 - a binary operation (`<<`, `>>`...) where either the left and/or right operator is doing on the same level a non binary operation (`*`, `+`, `/`...)
 - or when a literal number is preceded on the same level by a `-` (minus operator) and followed by some method calls

-----

@camsteffen Aside from the lint name, this lint only lints on ambiguous case, I don't think `&&`, `||` or any other operator is ambiguous is his precedence. Linting on any other case is outside the scope of this PR.

**Comment by RalfJung** [comment]
> a binary operation (<<, >>...) where either the left and/or right expression is a non-binary operation (*, +, /...)

`+` is a binary operator. "binary operator" means "operator with two arguments".

I had to read your statement 3 times until I realized you meant bitwise operator (as opposed to arithmetic operator).

**Comment by RalfJung** [comment]
So, for nomination...

my main case in point is that I wrote [this code](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=dd36361aa3ba6d12bc0a1295e50f8b5c) and then assumed that `ln_gamma` is just horribly imprecise because it gave results that are far away from it spec. The key expression is `-0.5f32.gamma()`. It never occurred to me that this might be parsed as `-(0.5f32.gamma())` -- *even though in the file I was doing this in there are plenty of `(-1.0f32).method`*, indicating that I ran into this in the past and then forgot again.

This is, of course, only anecdotal evidence. But it shows that even someone who is experienced in Rust, PL theory, and mathematics, can still be led completely astray by the parser behavior. There also is a clear way to disambiguate, so IMO this is a clear case for a warn-by-default lint.

**Comment by RalfJung** [comment]
> This lint is a bit slippery to define because there is a sliding scale of how much a human might know the precedence rules. 

Agreed. That's why it is a lint, not a hard error of any sort. We'll have to make judgment calls on which cases are sufficiently surprising that we want to lint about them by default. To me, `-5.method()` definitely passes that bar. `a && b || c` is near the edge -- I would generally reject such code during review and demand clarifying parentheses. `a + b*c` is fine since "multiplication binds stronger than addition" is extremely widely accepted.

File: src/tools/clippy/clippy_lints/src/renamed_lints.rs [hunk 2]
```diff
@@ -52,6 +52,7 @@ pub static RENAMED_LINTS: &[(&str, &str)] = &[
     ("clippy::invalid_utf8_in_unchecked", "invalid_from_utf8_unchecked"),
     ("clippy::let_underscore_drop", "let_underscore_drop"),
     ("clippy::mem_discriminant_non_enum", "enum_intrinsics_non_enums"),
+    ("clippy::precedence", "ambiguous_precedence"),
```
Comment (flip1995) [inline src/tools/clippy/clippy_lints/src/renamed_lints.rs:55]: You'll also have to add tests to `src/tools/clippy/tests/ui/renamed.rs`

**Comment by camsteffen** [comment]
> `-5.method()`

Hmm I take back my idea about "uncommon operators".

Maybe I'm being too pedantic but I don't like the name `ambiguous_precedence` because it never is actually ambiguous, only mistakenly understood by human readers... maybe `unclear_precedence`.

**Comment by RalfJung** [comment]
Note that lint names should be in plural form, e.g. "unclear_precedences".

**Comment by traviscross** [comment]
> Maybe I'm being too pedantic but I don't like the name `ambiguous_precedence` because it never is actually ambiguous, only mistakenly understood by human readers... maybe `unclear_precedence`.

If the focus is on the human, I'd go with `potentially_suprising_precedences`.

**Comment by blyxyas** [comment]
If Rust is a language meant to be written and reviewed by humans, I think the focus would always be on the human :sweat_smile:. 

imo `potentially_surprising_precedences` is too long, and the word "ambiguous" means that something has two meanings, which does not ever happen in the language as the syntax is deterministc. So, there's a +1 from me to `unclear_precedences`

**Comment by traviscross** [comment]
The word `unclear` has the same problem as `ambiguous`.  The precedence is *clear*.  There's nothing unclear about it.  It's just surprising to some humans.  If length is a concern (and we otherwise wanted this lint), we could say just `surprising_precedences`.  Or facetiously, we could say, `huh_precedences` or `wow_precedences`.  Those would be shorter.

**Comment by RalfJung** [comment]
"clear"/"unambiguous" is also often used to mean "obvious to the reader, does not require further context". By that standard, the precedence is not clear: I can't just show the code to someone and they will know what it means; they have to first learn the rules. After all we also talk about APIs being unclear when they use poorly chosen names or things like that, even though in a technical sense everything is well-defined and hence "clear".

**Pull request closed without merging**